cmd/hook.go

```diff
@@ -66,6 +66,7 @@ func runHookPreReceive(c *cli.Context) error {
 	reponame := os.Getenv(models.EnvRepoName)
 	userID, _ := strconv.ParseInt(os.Getenv(models.EnvPusherID), 10, 64)
 	prID, _ := strconv.ParseInt(os.Getenv(models.ProtectedBranchPRID), 10, 64)
+	isDeployKey, _ := strconv.ParseBool(os.Getenv(models.EnvIsDeployKey))
```

Comment:
As this comes from the environment, there's a risk it's spoofed. This value should be calculated in-process from the key itself.

Reply:
It's set in serv.go

Reply:
Apologies if I'm misunderstanding your comment - but I'll quickly run through how git and git hook work:

* Say you do a `git push` to any SSH git server. Once it is ready, `git` will effectively send `ssh git@giteaserver "git receive-pack :username/:reponame"` - the actual data will be sent over stdin / stdout for that connection.
* Now assuming you're logging in with a key, ssh will match your key with a line in authorized_keys (or it uses an AuthorizedKeys command to generate the line.)
* In Gitea's case this will match something like: 
```
command="/home/git/gitea/gitea serv key-2 --config='/home/git/gitea/custom/conf/app.ini'",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty ssh-rsa AAAAB3Nz... user@something
```
* When SSH sees the `command="..."` it immediately runs that command and stores a copy of the command that it was told to run in `$SSH_ORIGINAL_COMMAND`
* Note that `key-2` - SSH *never* exposes the key to command or to SSH_ORIGINAL_COMMAND if it runs.
  * In the case of the internal SSH it's basically the same.
* `gitea serv` needs to ensure that `$SSH_ORIGINAL_COMMAND` is one of the acceptable commands so if it's not it will stop there.
* `gitea serv` will check if the key argument provided should have access to the repository at all - but because it doesn't know what you're sending it cannot assess if you're trying to access a protected branch - it also cannot know if the commit fails. It therefore interprets the key-2 and expands it to a set of environment variables see: https://github.com/go-gitea/gitea/blob/f9aba9ba0f07b77cb46dde6eda3c3f5b8fa841fe/cmd/serv.go#L187-L193
  * `gitea serv` is, now, mostly a lightweight shim that does almost of all its work by sending a request to `/private/serv/command/:keyid/:owner/:repo` - however it does need to know where gitea is etc. hence the `--config`.
  * It's also worth noting that we're running `gitea serv` here as a new process - in a possibly completely separate environment to that which starts the `gitea web` process hence my comments on environment-to-ini PR - it can only configure by the ini - so the ini has to be correct.
* Once it's set the environment variables it will then exec the correct git command.
* It is the git commands, within its environment, that will then call the hooks that are stored in the git repository e.g. `.git/hooks/pre-receive.d/gitea`. So environment variables are the only way to configure this.
  * It's only during pre-receive that you can actually work out if you're touching a protected branch.
  * The calling of hooks explains why if people change their gitea command name they have to re-synchronize hooks - (there is another solution to this I've spoken about before.)

Unless you want to rewrite git entirely you only have environment variables but you do have control over these.

Reply:
I'm sorry, as I was looking at the small diff window I got confused with cmd serv, where we do check those things. My apologies.

Reply:
No worries.